### PR TITLE
Introduce log reporter and queue mechanism for event publishing

### PR DESCRIPTION
**Pull request by tishan89**
## Purpose
- To address https://github.com/wso2/apim-analytics-publisher/issues/17
- To provide an easy way to test intrumentation

## Goals
- Adds event queuing capability
- Adds a Log Metric Reporter which will output metric events to the log

## Approach
- Queuing is implemented using a bounded ArrayBlockingQueue. Producers will drop events if the queue is full. Consumers will keep on consuming the queue till queue becomes empty. EventHub client batching is also enabled
- Each metric attribute map will be converted to string and logged

## Documentation
To use log metric reporter
```
MetricReporter metricReporter = MetricReporterFactory.getInstance().createMetricReporter(
                "org.wso2.am.analytics.publisher.reporter.log.LogMetricReporter", null);
        CounterMetric metric = metricReporter.createCounterMetric("testCounter", null);

        Map<String, String> event = new HashMap<>();
        event.put("attribute1", "value1");
        event.put("attribute2", "value2");
        event.put("attribute3", "value3");
        event.put("attribute4", "value4");

        metric.incrementCount(event);
```

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/choreo/ChoreoAnalyticsThreadFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (RAVEENSR) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/choreo/ChoreoAnalyticsThreadFactory.java:2]:
```suggestion
 * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/log/LogCounterMetric.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class LogCounterMetric implements CounterMetric {
+    private static final Logger log = Logger.getLogger(LogCounterMetric.class);
+    private final String name;
+
+    protected LogCounterMetric(String name) {
+        this.name = name;
+    }
+
+    @Override public int incrementCount(Map<String, String> properties) throws MetricReportingException {
```
Comment (ruks) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/log/LogCounterMetric.java:39]: no need a new line after the annotation?

File: component/publisher-client/src/test/java/org/wso2/am/analytics/publisher/LogMetricReporterTestCase.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                "org.wso2.am.analytics.publisher.reporter.log.LogMetricReporter", null);
+        CounterMetric metric = metricReporter.createCounterMetric("testCounter", null);
+
+        Map<String, String> event = new HashMap<>();
+        event.put("attribute1", "value1");
+        event.put("attribute2", "value2");
+        event.put("attribute3", "value3");
+        event.put("attribute4", "value4");
+
+        metric.incrementCount(event);
```
Comment (fazlan-nazeem) [inline component/publisher-client/src/test/java/org/wso2/am/analytics/publisher/LogMetricReporterTestCase.java:49]: Shouldn't we assert something here?

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/log/LogCounterMetric.java (hunk 2 above)
Comment (tishan89) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/log/LogCounterMetric.java]: Checkstyles didnt catch. But anyway fixed with ae699e7e3b6108f6396fb3b470d5dda7ff226233

File: component/publisher-client/src/test/java/org/wso2/am/analytics/publisher/LogMetricReporterTestCase.java (hunk 3 above)
Comment (tishan89) [reply component/publisher-client/src/test/java/org/wso2/am/analytics/publisher/LogMetricReporterTestCase.java:49]: Currently we are depending on the fact that no exceptions are occurring. But latter will add a log appender and assert the log. Just didnt do it coz of the pending builder changes

**Review by ruks — APPROVED** [review]

**Pull request merged**